File: Src/DotNet/MobileNet/Binding/Binding.cs
/*
Copyright (c) 2012-2015, Coherent Labs AD
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of Coherent UI nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
using System;
using System.Runtime.InteropServices;

using Coherent.UI;
namespace Coherent.UI.Mobile.Binding
{
	/// <summary>
	/// Holds the arguments for the generic callback not handled by a registered delegate
	/// </summary>
	public class CallbackArguments
	{
		internal CallbackArguments(IntPtr data)
		{
            m_Data = data;

            int size = (int)Marshal.ReadInt64(data);
			var bytes = new byte[size];

			Marshal.Copy(data, bytes, 0, size);
			var reader = new Coherent.UI.Binding.BlobReader(bytes);
			var importer = new Coherent.UI.Binding.Importer(reader);
			Arguments = importer.ReadArguments();
		}

		/// <summary>
		/// Arguments provided by JavaScript
		/// </summary>
		public Value[] Arguments;

        internal IntPtr Data
        {
            get { return m_Data; }
        }

        private IntPtr m_Data;
	}

	/// <summary>
	/// Holds .Net extension methods to Coherent.UI.View
	/// </summary>
	public static class ViewExtensions
	{
		/// <summary>
		/// Trigger event in view
		/// </summary>
		/// <typeparam name="T1">Type for the event argument</typeparam>
		/// <param name="view">the view</param>
		/// <param name="name">name of the event</param>
		/// <param name="argument1">argument for the event</param>
		public static void TriggerEvent<T1>(this View view, string name, T1 argument1)
		{
			var exporter = view.Binder.GetExporter();

			exporter.Export(argument1);

			view.Binder.RawBinder.RawEvent(name, 1, exporter.GetBytesCount(), exporter.GetBytes());
		}

		/// <summary>
		///
		/// </summary>
		/// <typeparam name="T1">Type for the event argument</typeparam>
		/// <typeparam name="T2">Type for the event argument</typeparam>
		/// <param name="view">the view</param>
		/// <param name="name">name of the event</param>
		/// <param name="argument1">argument for the event</param>
		/// <param name="argument2">argument for the event</param>
		public static void TriggerEvent<T1, T2>(this View view, string name, T1 argument1, T2 argument2)
		{
			var exporter = view.Binder.GetExporter();

			exporter.Export(argument1);
			exporter.Export(argument2);

			view.Binder.RawBinder.RawEvent(name, 2, exporter.GetBytesCount(), exporter.GetBytes());
		}

		/// <summary>
		///
		/// </summary>
		/// <typeparam name="T1">Type for the event argument</typeparam>
		/// <typeparam name="T2">Type for the event argument</typeparam>
		/// <typeparam name="T3">Type for the event argument</typeparam>
		/// <param name="view">the view</param>
		/// <param name="name">name of the event</param>
		/// <param name="argument1">argument for the event</param>
		/// <param name="argument2">argument for the event</param>
		/// <param name="argument3">argument for the event</param>
		public static void TriggerEvent<T1, T2, T3>(this View view, string name, T1 argument1, T2 argument2, T3 argument3)
		{
			var exporter = view.Binder.GetExporter();

			exporter.Export(argument1);
			exporter.Export(argument2);
			exporter.Export(argument3);

			view.Binder.RawBinder.RawEvent(name, 3, exporter.GetBytesCount(), exporter.GetBytes());
		}

		/// <summary>
		///
		/// </summary>
		/// <typeparam name="T1">Type for the event argument</typeparam>
		/// <typeparam name="T2">Type for the event argument</typeparam>
		/// <typeparam name="T3">Type for the event argument</typeparam>
		/// <typeparam name="T4">Type for the event argument</typeparam>
		/// <param name="view">the view</param>
		/// <param name="name">name of the event</param>
		/// <param name="argument1">argument for the event</param>
		/// <param name="argument2">argument for the event</param>
		/// <param name="argument3">argument for the event</param>
		/// <param name="argument4">argument for the event</param>
		public static void TriggerEvent<T1, T2, T3, T4>(this View view, string name, T1 argument1, T2 argument2, T3 argument3, T4 argument4)
		{
			var exporter = view.Binder.GetExporter();

			exporter.Export(argument1);
			exporter.Export(argument2);
			exporter.Export(argument3);
			exporter.Export(argument4);

			view.Binder.RawBinder.RawEvent(name, 4, exporter.GetBytesCount(), exporter.GetBytes());
		}

		/// <summary>
		///
		/// </summary>
		/// <typeparam name="T1">Type for the event argument</typeparam>
		/// <typeparam name="T2">Type for the event argument</typeparam>
		/// <typeparam name="T3">Type for the event argument</typeparam>
		/// <typeparam name="T4">Type for the event argument</typeparam>
		/// <typeparam name="T5">Type for the event argument</typeparam>
		/// <param name="view">the view</param>
		/// <param name="name">name of the event</param>
		/// <param name="argument1">argument for the event</param>
		/// <param name="argument2">argument for the event</param>
		/// <param name="argument3">argument for the event</param>
		/// <param name="argument4">argument for the event</param>
		/// <param name="argument5">argument for the event</param>
		public static void TriggerEvent<T1, T2, T3, T4, T5>(this View view, string name, T1 argument1, T2 argument2, T3 argument3, T4 argument4, T5 argument5)
		{
			var exporter = view.Binder.GetExporter();

			exporter.Export(argument1);
			exporter.Export(argument2);
			exporter.Export(argument3);
			exporter.Export(argument4);
			exporter.Export(argument5);

			view.Binder.RawBinder.RawEvent(name, 5, exporter.GetBytesCount(), exporter.GetBytes());
		}

		/// <summary>
		///
		/// </summary>
		/// <typeparam name="T1">Type for the event argument</typeparam>
		/// <typeparam name="T2">Type for the event argument</typeparam>
		/// <typeparam name="T3">Type for the event argument</typeparam>
		/// <typeparam name="T4">Type for the event argument</typeparam>
		/// <typeparam name="T5">Type for the event argument</typeparam>
		/// <typeparam name="T6">Type for the event argument</typeparam>
		/// <param name="view">the view</param>
		/// <param name="name">name of the event</param>
		/// <param name="argument1">argument for the event</param>
		/// <param name="argument2">argument for the event</param>
		/// <param name="argument3">argument for the event</param>
		/// <param name="argument4">argument for the event</param>
		/// <param name="argument5">argument for the event</param>
		/// <param name="argument6">argument for the event</param>
		public static void TriggerEvent<T1, T2, T3, T4, T5, T6>(this View view, string name, T1 argument1, T2 argument2, T3 argument3, T4 argument4, T5 argument5, T6 argument6)
		{
			var exporter = view.Binder.GetExporter();

			exporter.Export(argument1);
			exporter.Export(argument2);
			exporter.Export(argument3);
			exporter.Export(argument4);
			exporter.Export(argument5);
			exporter.Export(argument6);

			view.Binder.RawBinder.RawEvent(name, 6, exporter.GetBytesCount(), exporter.GetBytes());
		}

		/// <summary>
		///
		/// </summary>
		/// <typeparam name="T1">Type for the event argument</typeparam>
		/// <typeparam name="T2">Type for the event argument</typeparam>
		/// <typeparam name="T3">Type for the event argument</typeparam>
		/// <typeparam name="T4">Type for the event argument</typeparam>
		/// <typeparam name="T5">Type for the event argument</typeparam>
		/// <typeparam name="T6">Type for the event argument</typeparam>
		/// <typeparam name="T7">Type for the event argument</typeparam>
		/// <param name="view">the view</param>
		/// <param name="name">name of the event</param>
		/// <param name="argument1">argument for the event</param>
		/// <param name="argument2">argument for the event</param>
		/// <param name="argument3">argument for the event</param>
		/// <param name="argument4">argument for the event</param>
		/// <param name="argument5">argument for the event</param>
		/// <param name="argument6">argument for the event</param>
		/// <param name="argument7">argument for the event</param>
		public static void TriggerEvent<T1, T2, T3, T4, T5, T6, T7>(this View view, string name, T1 argument1, T2 argument2, T3 argument3, T4 argument4, T5 argument5, T6 argument6, T7 argument7)
		{
			var exporter = view.Binder.GetExporter();

			exporter.Export(argument1);
			exporter.Export(argument2);
			exporter.Export(argument3);
			exporter.Export(argument4);
			exporter.Export(argument5);
			exporter.Export(argument6);
			exporter.Export(argument7);

			view.Binder.RawBinder.RawEvent(name, 7, exporter.GetBytesCount(), exporter.GetBytes());
		}

		/// <summary>
		///
		/// </summary>
		/// <typeparam name="T1">Type for the event argument</typeparam>
		/// <typeparam name="T2">Type for the event argument</typeparam>
		/// <typeparam name="T3">Type for the event argument</typeparam>
		/// <typeparam name="T4">Type for the event argument</typeparam>
		/// <typeparam name="T5">Type for the event argument</typeparam>
		/// <typeparam name="T6">Type for the event argument</typeparam>
		/// <typeparam name="T7">Type for the event argument</typeparam>
		/// <typeparam name="T8">Type for the event argument</typeparam>
		/// <param name="view">the view</param>
		/// <param name="name">name of the event</param>
		/// <param name="argument1">argument for the event</param>
		/// <param name="argument2">argument for the event</param>
		/// <param name="argument3">argument for the event</param>
		/// <param name="argument4">argument for the event</param>
		/// <param name="argument5">argument for the event</param>
		/// <param name="argument6">argument for the event</param>
		/// <param name="argument7">argument for the event</param>
		/// <param name="argument8">argument for the event</param>
		public static void TriggerEvent<T1, T2, T3, T4, T5, T6, T7, T8>(this View view, string name, T1 argument1, T2 argument2, T3 argument3, T4 argument4, T5 argument5, T6 argument6, T7 argument7, T8 argument8)
		{
			var exporter = view.Binder.GetExporter();

			exporter.Export(argument1);
			exporter.Export(argument2);
			exporter.Export(argument3);
			exporter.Export(argument4);
			exporter.Export(argument5);
			exporter.Export(argument6);
			exporter.Export(argument7);
			exporter.Export(argument8);

			view.Binder.RawBinder.RawEvent(name, 8, exporter.GetBytesCount(), exporter.GetBytes());
		}
	}
}

File: Src/Binding/Binding.cpp
/*
Copyright (c) 2012-2015, Coherent Labs AD
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of Coherent UI nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
#include "precompiled.h"

#include <Binding/Binding.h>
#include <Binding/Property.h>
#include <Binding/Detail/TypeInfo.h>

namespace Coherent
{
namespace UI
{

class BinderImpl
{
public:
	BinderImpl()
		: m_CurrentObject(nullptr)
	{
	}

	TypeDescription RegisterType(const char* name, void* object)
	{
		if (!m_CurrentObject)
		{
			return (BeginType(name, object))? TypeDescription(m_Public, object) : TypeDescription(nullptr, nullptr);
		}
		else if (m_CurrentObject != object)
		{
			ptrdiff_t offset = reinterpret_cast<size_t>(object) - reinterpret_cast<size_t>(m_CurrentObject);
			assert(offset == (int)offset);
			m_CurrentType->AddProperty(new AdjustPointer(name, (int)offset));
			m_CurrentObject = object;
		}
		return TypeDescription(m_Public, object);
	}

	void AddProperty(void* object, const Property& property)
	{
		if (m_CurrentObject != object)
		{
			ptrdiff_t offset = reinterpret_cast<size_t>(object) - reinterpret_cast<size_t>(m_CurrentObject);
			assert(offset == (int)offset);
			m_CurrentType->AddProperty(new AdjustPointer(nullptr, (int)offset));
			m_CurrentObject = object;
		}
		m_CurrentType->AddProperty(property.Clone());
	}

	void SetPublic(Binder* binder)
	{
		m_Public = binder;
	}

	bool BeginType(const char* name, void* object)
	{
		// refuse to register a type via nullptr because a dynamic_cast (or some virtual inheritance)
		// in user code will result in a crash
		if (object == nullptr)
		{
			if (m_IsBinding)
			{
				m_Public->BindNull();
			}
			return false;
		}
		// TODO: remove the temporary std::string(name)
#if !defined(COHERENT_COMPILER_GXX)
		auto newType = m_TypeData.emplace(TypeData::value_type(name, TypeInfo()));
#else
		auto newType = m_TypeData.insert(TypeData::value_type(name, TypeInfo()));
#endif
		if (newType.second)
		{
			m_CurrentObject = object;
			m_CurrentType = &newType.first->second;
			m_CurrentType->SetName(name);
			return true;
		}

		if(m_IsBinding)
		{
			newType.first->second.Bind(m_Public, object);
		}
		else
		{
			newType.first->second.Read(m_Public, object);
		}

		return false;
	}

	void SetBinding(bool isBinding)
	{
		m_IsBinding = isBinding;
		m_CurrentObject = nullptr;
		m_CurrentType = nullptr;
	}
	
	void EndType(void* object)
	{
		if (m_CurrentType)
		{
			if (m_IsBinding)
			{
				m_CurrentType->Bind(m_Public, object);
			}
			else
			{
				m_CurrentType->Read(m_Public, object);
			}
			m_CurrentType = nullptr;
		}
	}

private:
	BinderImpl& operator=(const BinderImpl&) = delete;
	
	// TODO: remove this O(log(N)) look up
	typedef std::map<std::string, TypeInfo, std::less<std::string>> TypeData;
	
	TypeData m_TypeData;

	void* m_CurrentObject;
	TypeInfo* m_CurrentType;
	
	Binder* m_Public;

	bool m_IsBinding;
};

TypeDescription::TypeDescription(Binder* binder, void* object)
	: m_Binder(binder)
	, m_Object(object)
{
}

TypeDescription::TypeDescription(const TypeDescription& other)
	: m_Binder(other.m_Binder)
	, m_Object(other.m_Object)
{
}

TypeDescription::~TypeDescription()
{
}

Binder::Binder()
	: m_Impl(new BinderImpl)
{
	m_Impl->SetPublic(this);
}


Binder::~Binder()
{
	delete m_Impl;
}

TypeDescription Binder::RegisterType(const char* name, void* object)
{
	return m_Impl->RegisterType(name, object);
}

TypeDescription Binder::RegisterType(const char* name, const void* object)
{
	return m_Impl->RegisterType(name, const_cast<void*>(object));
}

void Binder::AddProperty(void* object, const Property& property)
{
	m_Impl->AddProperty(object, property);
}

void Binder::BeginType()
{
	m_Impl->SetBinding(true);
}

void Binder::ReadType()
{
	m_Impl->SetBinding(false);
}

void Binder::EndType(void* object)
{
	m_Impl->EndType(object);
}

void CoherentBindInternal(Binder* binder, bool value)
{
	binder->Bind(value);
}

void CoherentBindInternal(Binder* binder, signed int value)
{
	binder->Bind(value);
}

void CoherentBindInternal(Binder* binder, unsigned int value)
{
	binder->Bind(value);
}

void CoherentBindInternal(Binder* binder, signed char value)
{
	binder->Bind(value);
}

void CoherentBindInternal(Binder* binder, unsigned char value)
{
	binder->Bind(value);
}

void CoherentBindInternal(Binder* binder, float value)
{
	binder->Bind(value);
}

void CoherentBindInternal(Binder* binder, double value)
{
	binder->Bind(value);
}

void CoherentBindInternal(Binder* binder, const char* value)
{
	binder->Bind(value);
}


void CoherentBindInternal(Binder* binder, const wchar_t* value)
{
	binder->Bind(value);
}

void CoherentReadInternal(Binder* binder, bool& value)
{
	binder->Read(value);
}

void CoherentReadInternal(Binder* binder, signed int& value)
{
	binder->Read(value);
}

void CoherentReadInternal(Binder* binder, unsigned int& value)
{
	binder->Read(value);
}

void CoherentReadInternal(Binder* binder, signed char& value)
{
	binder->Read(value);
}

void CoherentReadInternal(Binder* binder, unsigned char& value)
{
	binder->Read(value);
}

void CoherentReadInternal(Binder* binder, float& value)
{
	binder->Read(value);
}

void CoherentReadInternal(Binder* binder, double& value)
{
	binder->Read(value);
}

}
}
